File: unrated/typical90_q2.py
n, m = map(int, input().split())
lr = [list(map(int, input().split())) for _ in range(m)]

if n > 1000:
    exit()

# 小課題2 O(n2)
out = [[0] * n for i in range(n)]
for l, r in lr:
    out[l-1][r-1] = 1

ans = 0
lst = [0] * n
for i in range(n):
    sumout, lst[i] = sum(out[i]), 0
    for j in range(n):
        sumout -= out[i][j]
        ans += sumout * lst[j]
        lst[j] += out[i][j]
print(ans)

File: unrated/typical90_q2.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using matll = vector<vector<ll>>;

#define sum(v) accumulate(v.begin(), v.end(), 0LL)
#define rep(i, n) for (ll i = 0; i < ll(n); i++)
#define for_itr(id, itr) for (auto& id : itr)

//小課題2 O(n2)
int main() {
    ll n, m;
    cin >> n >> m;
    matll lr(m, vector<ll>(2, -1));
    rep(i, m) cin >> lr[i][0] >> lr[i][1];

    if (n > 1000) return 0;
    matll out(n, vector<ll>(n, 0));
    for_itr(lri, lr) out[lri[0] - 1][lri[1] - 1]++;

    ll ans = 0;
    vector<ll> vec(n, 0);
    rep(i, n) {
        ll sumout = sum(out[i]);
        vec[i] = 0;
        rep(j, n) {
            sumout -= out[i][j];
            ans += sumout * vec[j];
            vec[j] += out[i][j];
        }
    }
    cout << ans << endl;
    return 0;
}
